File: lib/src/rules/predicate.dart
// lib/src/rules/predicate.dart
class PredicateFormatException implements Exception {
  final String message;
  PredicateFormatException(this.message);
  @override
  String toString() => 'PredicateFormatException: $message';
}

class GameState {
  final Set<String> carvedFragmentIds;
  /// Canonical form: two entity ids sorted and joined with '|'.
  final Set<String> linkedPairs;
  final Set<String> answeredQuestionIds;

  const GameState({
    required this.carvedFragmentIds,
    required this.linkedPairs,
    required this.answeredQuestionIds,
  });

  static String linkKey(String a, String b) {
    final pair = [a, b]..sort();
    return '${pair[0]}|${pair[1]}';
  }

  bool hasLink(String a, String b) =>
      linkedPairs.contains(linkKey(a, b)) || linkedPairs.contains('$a|$b') ||
      linkedPairs.contains('$b|$a');
}

abstract class Predicate {
  const Predicate();
  bool evaluate(GameState s);
}

class CarvedPredicate extends Predicate {
  final String fragmentId;
  const CarvedPredicate(this.fragmentId);
  @override
  bool evaluate(GameState s) => s.carvedFragmentIds.contains(fragmentId);
}

class LinkedPredicate extends Predicate {
  final String a;
  final String b;
  const LinkedPredicate(this.a, this.b);
  @override
  bool evaluate(GameState s) => s.hasLink(a, b);
}

class AnsweredPredicate extends Predicate {
  final String questionId;
  const AnsweredPredicate(this.questionId);
  @override
  bool evaluate(GameState s) => s.answeredQuestionIds.contains(questionId);
}

class AllPredicate extends Predicate {
  final List<Predicate> children;
  const AllPredicate(this.children);
  @override
  bool evaluate(GameState s) => children.every((c) => c.evaluate(s));
}

class AnyPredicate extends Predicate {
  final List<Predicate> children;
  const AnyPredicate(this.children);
  @override
  bool evaluate(GameState s) => children.any((c) => c.evaluate(s));
}

class NotPredicate extends Predicate {
  final Predicate child;
  const NotPredicate(this.child);
  @override
  bool evaluate(GameState s) => !child.evaluate(s);
}

/// The complete grammar. Adding a key requires a decision record in the design spec.
const _allowedKeys = {'carved', 'linked', 'answered', 'all', 'any', 'not'};

Predicate parsePredicate(Map<String, dynamic> json) {
  if (json.length != 1) {
    throw PredicateFormatException(
        'A predicate must have exactly one key, got ${json.keys.toList()}');
  }
  final key = json.keys.first;
  if (!_allowedKeys.contains(key)) {
    throw PredicateFormatException(
        'Unknown predicate "$key". Allowed: ${_allowedKeys.join(', ')}. '
        'Expression strings and script references are not supported (INV-5).');
  }
  final value = json[key];

  switch (key) {
    case 'carved':
      if (value is! String) {
        throw PredicateFormatException('"carved" needs a fragment id string');
      }
      return CarvedPredicate(value);
    case 'answered':
      if (value is! String) {
        throw PredicateFormatException('"answered" needs a question id string');
      }
      return AnsweredPredicate(value);
    case 'linked':
      if (value is! List || value.length != 2) {
        throw PredicateFormatException('"linked" needs exactly two entity ids');
      }
      return LinkedPredicate(value[0] as String, value[1] as String);
    case 'all':
    case 'any':
      if (value is! List || value.isEmpty) {
        throw PredicateFormatException('"$key" needs a non-empty list');
      }
      final children = value
          .map((c) => parsePredicate(Map<String, dynamic>.from(c as Map)))
          .toList();
      return key == 'all' ? AllPredicate(children) : AnyPredicate(children);
    case 'not':
      if (value is! Map) {
        throw PredicateFormatException('"not" needs a predicate object');
      }
      return NotPredicate(parsePredicate(Map<String, dynamic>.from(value)));
    default:
      throw PredicateFormatException('Unreachable');
  }
}

File: Sources/CarveCore/Rules/Predicate.swift
public struct PredicateFormatError: Error, Equatable, CustomStringConvertible {
  public let message: String
  public init(_ message: String) {
    self.message = message
  }
  public var description: String { message }
}

public struct GameState: Equatable, Sendable {
  public let carvedFragmentIds: Set<String>
  /// Canonical form: two entity ids sorted and joined with '|'.
  public let linkedPairs: Set<String>
  public let answeredQuestionIds: Set<String>

  public static func linkKey(_ a: String, _ b: String) -> String {
    let pair = [a, b].sorted()
    return "\(pair[0])|\(pair[1])"
  }

  // FIX A: canonical form only. The read path must accept exactly what
  // `linkKey` produces; raw orderings are never valid, or the
  // canonicalization contract becomes unenforceable.
  public func hasLink(_ a: String, _ b: String) -> Bool {
    linkedPairs.contains(Self.linkKey(a, b))
  }
}

public protocol Predicate: Sendable {
  func evaluate(_ state: GameState) -> Bool
}

public struct CarvedPredicate: Predicate {
  public let fragmentId: String
  public func evaluate(_ state: GameState) -> Bool {
    state.carvedFragmentIds.contains(fragmentId)
  }
}

public struct LinkedPredicate: Predicate {
  public let a: String
  public let b: String
  public func evaluate(_ state: GameState) -> Bool {
    state.hasLink(a, b)
  }
}

public struct AnsweredPredicate: Predicate {
  public let questionId: String
  public func evaluate(_ state: GameState) -> Bool {
    state.answeredQuestionIds.contains(questionId)
  }
}

public struct AllPredicate: Predicate {
  public let children: [any Predicate]
  public func evaluate(_ state: GameState) -> Bool {
    children.allSatisfy { $0.evaluate(state) }
  }
}

public struct AnyPredicate: Predicate {
  public let children: [any Predicate]
  public func evaluate(_ state: GameState) -> Bool {
    children.contains { $0.evaluate(state) }
  }
}

public struct NotPredicate: Predicate {
  public let child: any Predicate
  public func evaluate(_ state: GameState) -> Bool {
    !child.evaluate(state)
  }
}

/// The complete grammar. Adding a key requires a decision record in the design
/// spec (INV-5).
private let allowedPredicateKeys: Set<String> = [
  "carved", "linked", "answered", "all", "any", "not",
]

// FIX B: entirely cast-free. `guard case .string(...) = value` pattern matching
// only — a malformed element throws `PredicateFormatError`, never crashes.
public func parsePredicate(_ object: [String: JSONValue]) throws -> any Predicate {
  guard let (key, value) = object.first, object.count == 1 else {
    throw PredicateFormatError("A predicate must have exactly one key, got \(Array(object.keys))")
  }
  guard allowedPredicateKeys.contains(key) else {
    throw PredicateFormatError(
      "Unknown predicate \"\(key)\". Allowed: carved, linked, answered, all, any, not. "
        + "Expression strings and script references are not supported (INV-5).")
  }

  switch key {
  case "carved":
    guard case .string(let fragmentId) = value else {
      throw PredicateFormatError("\"carved\" needs a fragment id string")
    }
    return CarvedPredicate(fragmentId: fragmentId)
  case "answered":
    guard case .string(let questionId) = value else {
      throw PredicateFormatError("\"answered\" needs a question id string")
    }
    return AnsweredPredicate(questionId: questionId)
  case "linked":
    guard case .array(let elements) = value,
      elements.count == 2,
      case .string(let a) = elements[0],
      case .string(let b) = elements[1]
    else {
      throw PredicateFormatError("\"linked\" needs exactly two entity ids")
    }
    return LinkedPredicate(a: a, b: b)
  case "all", "any":
    guard case .array(let elements) = value, !elements.isEmpty else {
      throw PredicateFormatError("\"\(key)\" needs a non-empty list")
    }
    var children: [any Predicate] = []
    children.reserveCapacity(elements.count)
    for element in elements {
      guard case .object(let childObject) = element else {
        throw PredicateFormatError("\"\(key)\" children must be predicate objects")
      }
      children.append(try parsePredicate(childObject))
    }
    if key == "all" {
      return AllPredicate(children: children)
    }
    return AnyPredicate(children: children)
  case "not":
    guard case .object(let childObject) = value else {
      throw PredicateFormatError("\"not\" needs a predicate object")
    }
    return NotPredicate(child: try parsePredicate(childObject))
  default:
    throw PredicateFormatError("Unreachable")
  }
}
